### ISSUE TITLE
Multiple users in one task

### ISSUE THREAD

**Issue by willianwrm** [kanboard/kanboard#85]
Sometimes a task is done by more than one person, it would be nice to add such feature.
Thanks in advance.

**Comment by gmdesousa** [comment]
I agree with the possibility to assign two people to one task?
Is this feature going to be implemented?

**Comment by rafasim** [comment]
My team works with other companies and for various activities we put team members as supervisors of the activities carried out by an employee of the third company. In this case, an activity have multiple users can be extremely useful.

**Comment by filipesaraiva** [comment]
It would be a really usual feature. +1.

**Comment by cknoll** [comment]
I really want that feature, too. And our group would need it asap.

Maybe those people who want it, can underline their whish with the help of some platform like
https://www.bountysource.com/ ("Get features added and bugs fixed")
https://freedomsponsors.org/ ("Crowdfunding Free Software, one issue at time")

I have no experience with such stuff, and maybe it would distort the development process if money comes into such projects. But on the other hand I can imagine things people spend money for, which are much worse than new features in cool pieces of software.

**Comment by nekohayo** [comment]
@cknoll (or anyone else) can't someone in your group provide a patch for this instead? Typically better than throwing money at the problem (sometimes that doesn't change anything, the problem is usually time) or pressuring the developer.

**Comment by cknoll** [comment]
I fear, we are not sufficiently familiar with php to provide a patch. And I thought as the issue is open for more than 8 months, creating a bounty would communicate that there are people who are really interested in this feature.

**Comment by levlaz** [comment]
@fguillot I am thinking on how to achieve this. On the UI side we could just do a nice multi-select widget like this http://www.erichynds.com/examples/jquery-ui-multiselect-widget/demos/ and allow users to select multiple assignees. 

The problem in my opinion is with the data model. Currently it looks like the assignee is the task owner in the task table which is an integer value. We cannot place multiple integers into this table because that is nonsense. 

One solution would be to make an association table like task_assignee and have that resolve the Many to Many relationship between tasks and assignees. 

Another option would be to simply add a column such as "collaborators" into the tasks table and insert any user_id's into it as a CSV. We can them parse out the CSV list of user_id's and combine it with the owner_id to show all of the people that belong to this task. 

The second approach seems a bit hacky, but I would love to hear your thoughts. :)

**Comment by fguillot** [comment]
I'm ok to add this feature but I don't want to include any breaking changes.

So here a possible solution similar to the one above:

- A task can have several participants/collaborators/members in addition to the current assignee system
- Add a new table "task_has_members"
- Add a new shortcut menu to add quickly a new member to the task from the board
- From the task details, add a new page to add/remove members
- Display on the board the number of members for each task with a tooltip on mouseover that list all members
- Update board filters
- Update notifications

**Comment by fleadog99** [comment]
I like the possible solution you offered. Was there a status update on this request, or was this just an idea of how we could implement the change? This element is vital to my implementation at a production level.

**Comment by valessiobrito** [comment]
:+1:  This is a great feature.

**Comment by nekohayo** [comment]
Hi Frédéric, I think your proposed approach makes sense. At least it would fit the current workflow problem I'm trying to solve: a university research group, where students have an official/primary professor assigned as their research director, and then that professor can have co-directors and lieutenants (research associates under the professor that help mentor the students).

I think this could work in conjunction with issue #126

**Comment by Lesstat** [comment]
Hello togehter,
I thought that it would maybe help to change the filter for assigned to and include also all tasks where the User ist assigned to a subtask and (possibly) not to the task itself.

I imagine that this would be a smaller change to do and give a lot of the benefits that multiuser tasks have.  
What do you think?

**Comment by nekohayo** [comment]
I would like to humbly request this to be reopened/de-merged from issue #1012. I still see plenty of situations where I would want more than one assignee globally. Especially cards/projects that don't actually have/use subtasks (which is the case 99% of the time for the people I'm dealing with). There are cases out there in the real world where projects are assigned with "equal responsibility" to two people, and it would be very nice if it could have a visual avatar representation as suggested in issue #126.

**Comment by maurorivas2k** [comment]
I agree with this feature request, will be very usefull can assign to a tow o more resource at same time

**Comment by auguster** [comment]
I'd like to be able to assign a task to a group. Has project leader I don't really care who in particular handles the task.

Then the group can re-assign the task to one of its member.

**Comment by Minyun** [comment]
Assigning groups to tasks is vital to my process. When will this feature become available?

**Comment by paddymac841** [comment]
What sort of progress is being made on this issue? Has anyone attempted to develop this themselves? Any pointers to me would be appreciated. Have to implement this asap..

**Comment by levlaz** [comment]
@paddymac841 there looks like there has not been much  movement on this feature. 

>Have to implement this asap

If this is pretty urgent for your company then you can consider [sponsoring this feature](https://kanboard.net/consulting).

**Comment by fguillot** [comment]
The main reason that this feature is not yet developed is because the data model need to be modified. Actually the assignee is stored in the tasks table with a column `owner_id` (which is poorly named) and the best approach is probably to have a separate table for all assignees. That will require a lot of code refactoring, it's possible but takes time.

**Comment by ptman** [comment]
If more than one person are responsible, then in effect nobody is responsible. I think the current way is good, with one person responsible who is then able to create subtasks and assign them.

**Comment by Minyun** [comment]
I disagree... one should be able to assign any kind of task (parent or
child) to a group.

**Comment by ptman** [comment]
If it's a specific group instead of a freely chosen set of users you could always create users for mailing lists / groups. But it is only half of a solution.

**Comment by auguster** [comment]
For my usage, using kanboard groups for tasks would be great. I guess dynamic groups (or free groups) could be a appreciated feature but maybe harder to implement.

**Comment by nekohayo** [comment]
@ptman wrote:

> If more than one person are responsible, then in effect nobody is responsible. I think the current way is good, with one person responsible who is then able to create subtasks and assign them.

That's your perception based on how things work in _your_ organization(s), but no, that is not sufficient. As I pointed out in my previous comments, there are real world organizational scenarios out there where people have an equal shared responsibility without being part of a "team"; that's cross-functional/interdisciplinary assignments. In my case they are a small group (think: less than 10 people) that work in different pairs depending on each task. I can't create a "team" or "group" for every possible "pair", that's ridiculous, it doesn't help the users gain clarity.

**Comment by seancarson** [comment]
In my opinion, I believe subtasks or a dedicated project board work best for sharing a common scope of work with a team or group.  In my experience Tasks with subtasks work great for small items, say parent tasks with no more than 1-10 steps such as templates or repeatable processes.  From my experience anything over that needs to have its on project board.    

The feature referenced in #2411 would allow for group communication around a Parent Task with group/team members communicating status and updates with the group via comments which would be automatically emailed to all defined CC field on the task.  The ability to add an Kanboard Defined Group to an Task CC field would also accomplish the same thing as assigning a group to a task as well.    

A combination of a ParentTask\Subtask, Task CC field and the ability to add Kanboard Defined Groups to the Task CC should cover must use cases.  

SC

**Comment by NinjaPewPew** [comment]
I read all of the above, but I really don't understand why a simple sub-task with a specific assignee isn't enough. How would you be able to follow *who* done *what* in the task, if a "group" of some from is attached as responsible?

Even if its just as some sort of history, who done what, a sub-task created by each responsible (and assigned to them) is much better. Even if you have a lot of "pairs" in your company, as @nekohayo mentioned, I don't see how having 2 sub-task is some sort of "overkill". At the very least it would be clear in the end who did participate and who didn't (for example if they were sick or whatever, but belonged to the "pair").

**Comment by Minyun** [comment]
The bottom line is that there is a need for it. Even if this thread
produces a need that is 1 out of 10, the need exists. I suspect those that
need the change are not developer savvy and understand business need and
the developers are not very business savvy and do not see the point... As a
business, this is a requirement-please respect that.

**Comment by auguster** [comment]
From my point of view, multiple assignment at task level is a priority over multiple assignment at subtask level.

It's true that this is mostly for corporate aspects than for development (at least for now). When my organization grow, I expect to need this for dev teams also (though I'm not there yet ;).

**Comment by seancarson** [comment]
There is a way to do this right now in Kanboard. 

1. create a distro group in your mail system per team.  
2. Define a remote user with the distribution group email  and select disable login from form option.  
3. Assign said remote user to your board and begin assigning tasks to them.  Notification will then go to the users defined in the distribution group.  

Im using this in place of Kanboard Defined groups that have task assignment and notification ability.

**Comment by Minyun** [comment]
Nice one. Clean and simple.

**Comment by Philippe-Arnaud** [comment]
From my personal perspective, this changes the very core of the flow in Kanboard, breaking the beautiful underlying concepts of Kanban. A task should have one and only one people being in charge. If you feel that it's too restrictive, then the task is too big and should be spliced in smaller chunks.

**Comment by nekohayo** [comment]
@Philippe-Arnaud that's what works for *you* because you think of kanbard cards in terms of "tasks". It doesn't work in every organization when you think in terms of "projects". Please read my previous comment(s) on this thread.

**Comment by ptman** [comment]
A tool that tries to be everything to everyone isn't good at anything. We think of tasks as tasks because  they are created using a button labeled "Add a new task" and boards as projects since those are created using "New Project". At least I would try to avoid using a tool against its design.

**Comment by nekohayo** [comment]
> From my personal perspective, this changes the very core of the flow in Kanboard, breaking the beautiful underlying concepts of Kanban. [...]

I disagree, I strongly believe it doesn't.

> [...] We think of tasks as tasks because they are created using a button labeled "Add a new task" and boards as projects since those are created using "New Project". At least I would try to avoid using a tool against its design.

Try telling that to a manufacturing plant manager (y'know, where kanban board originated from). Or a medical clinic. Or a research lab. Or a very small consultancy group.

That the labelling of widgets on the kanboard UI does not fit reality is not a valid argument to me. I personally call kanboard's boards "boards", not "projects", because that's exactly what they are: boards. And the cards ("tasks") inside them are exactly what I'm saying: projects. And THEN those projects may or may not have subtasks, but that's up to the users.

There is *nothing* that you lose from having the optional ability to assign more than one person to a card. It's additive, not substractive, not a zero-sum game. Please don't impose your complexity/workflow/terminology onto everybody else. What you're doing is trying to force people to use this asphalted path: http://www.drewhajduk.co.uk/wp-content/uploads/2012/02/2847766967_70b6bec449_z.jpg

**Comment by krembo99** [comment]
I think that some people forget that the whole methodology is about **collaboration** of **teams** inside an **organization**  - not the single person's to-do-list. In many organizations, including mine , the tasks are collaborative by definition - and multidisciplinary  - and can never be performed by one single member, nor can be supervised by one for the simple reason that no one has the full set of skills required .  the team has the skill set , but not one individual. Assigning multiple users allows for example assigning two disciplines ( departments ) to collaborate on the same task ( via their heads .. )

**Comment by NinjaPewPew** [comment]
I think some people forget that even in an team environment, you need a way to organize the different members of that team. Every team member has different skills and not everybody can do every task. A task should be something very specific and contained in on Kanboard Task (aka Card). Every team member takes the task they can do or someone else assigns task to them.

That is the Kanban methodology.

**Comment by krembo99** [comment]
@NinjaPewPew  > That is the Kanban methodology.

I think that you are talking strictly from a software POV.
kanban was invented as an [inventory control system](https://en.wikipedia.org/wiki/Kanban) in large manufacturing plants . 
My organization is in fact, working with factories to install control systems for manufacturing and interfacing with real - life machinery and objects.
Tasks like " install a 5 axis cnc machine", or "upgrade 60XX placement robot " require hundreds of subtasks to be performed by a multidisciplinary team. 

They can not be broken into smaller tasks because of the following reasons :

1 - The teams and individuals already know their sub-tasks . There is no need to spell it out .
2 - You would need to hire someone ( or more ) just to assign tasks all day ( and also assign the task of assigning the tasks )  .
3 - where do you stop "individual" tasks ? at the machine level ? mechanism level ? screw level ?
4 -  You will end up producing maintanace manual on the board .
5  - and this is important  - you can not achieve same-time collaboration because each team member "picks up" their tasks when they are ready , they can, or they want ...

Overall, in simpler mechanisms ( read : software ) this approach might work (scrum)  . but in larger and more complex projects ( Architecture, engineering , manufacturing , to name a few  ) - multidisciplinary teams are de-facto.

Even simple tasks as " move the large table " can require 2 or 3 members . so what do you do ? assign it to the team leader , who then assign the task for each member individually ? you will end up assigning all task to the head of the organization for delegation - or in another case , all 3 team members report to moving the table at different times .

**Comment by lawrenceweetman** [comment]
I totally agree with the OP. Those who disagree and say that a "task" should only ever be taken on by one person don't sound as though they've worked in a truly collaborative team, where even on small tasks collaboration like at-keyboard pair programming takes place (as defined in XP).

**Comment by piltraspn** [comment]
I totally agree with everyone who thinks that Multiple asignees is absolutely necessary. Flexibility is a must today and sub-tasks are very limited. After almost 3 years, I think this would be an excellent addition to Kanboard.

**Comment by dhornford** [comment]
-1 you lose accountability and require one status. If you have true work teams create a user for the team

Dave Hornford

**Comment by lawrenceweetman** [comment]
@dhornford - It sounds as though you are only viewing Kanban boards through the lens of a board displaying multiple projects being worked on by multiple teams. 

If you're a single team (of, say, seven software developers) working on a single project then it's entirely feasible that you could have two people assigned to a task. For example, one writing tests and the other writing working code. Or better yet, pair programming: https://en.wikipedia.org/wiki/Pair_programming

If you still disagree, how do you recommend displaying two software engineers "pair programming" (super, super common these days) on a task on your Kanboard? Make pair-combinations for all 7 people on the team? 28 of them? Or just have each task represented by one of the two developers? What does that do for the Kanban principles of visualising workflow and having a transparent development cycle? Not use Kanboard?

At least it works for you, right?

**Comment by dhornford** [comment]
Your approach may vary

I do not see random combinations  as indicative of visualization of, or for, team. In fact would suggest that you might be de-visualizing.

If you have true teams that deliver (which we do, although we don't do any coding and our approach to Kanban is derived from experience manufacturing) you might want to consider team lanes or separate boards.

We use both, as our teams work in different geos & on different projects.

Have usually found lanes, and the analogy the a manufacturing line, far better at visualizing the team and the dashboard for individual review.

As before, your approach may vary. 

Dave Hornford

**Comment by lawrenceweetman** [comment]
@dhornford We do have separate boards. I'm specifically talking about a team board, where multiple team members work on one task. This happens in EVERY software team I know using any kind of board (Kanban, SCRUM, etc). 

Perhaps the issue here is that we work in different domains. In my domain, this feature is a necessity. In my domain, many people are trying to adopt Kanban. Your argument is for blocking the Software Engineering community from using Kanboard because the feature set is sufficient for your use case. That's a poor argument.

**Comment by dhornford** [comment]
As before your approach may vary

When we need parts of a task done by an individual we use a sub task & assignment, or a card with an internal link.

Hence, belief that feature set addresses need and work on this distracts from other enhancements.

Don't believe that my opinion more important than anyone else's in the developers' assessment of priority. For things really important to us we are funding development of a plugin.

All the best


Dave Hornford

**Comment by lawrenceweetman** [comment]
@dhornford 

That's all fine. Commercial Kanban tools generally support it. For me it was the deal breaker when I was evaluating this tool.

As you say, it's up to the devs. Given that it's open source it's completely within my prerogative to create a fix and offer it back to the community, but unfortunately I don't have time right now so I haven't been able to offer. 

All the best,
Lawrence

**Comment by Bronwenist** [comment]
@fguillot 

This feature would be very useful for us. If no one else is currently working on it, I'm interested in picking it up and having a go.

**Comment by tpimont** [comment]
![image](https://cloud.githubusercontent.com/assets/25609323/23651713/acde80be-0326-11e7-840d-a750ca50a64d.png)

Bonjour,
Cette fonctionnalité serait vraiment un plus.
Je veux bien y passer un peu de temps mais je ne sais pas par ou commencer .
Pourriez vous m'indiquer par ou débuter ?
Merci

**Comment by mapeze** [comment]
I would love too this feature.

**Comment by NoodleBB** [comment]
👍 for that feature. It's an elementary future.

**Comment by mrkafk** [comment]
This is a must-have really...

**Comment by Snajper88** [comment]
I'm in, too. We also need the possibility to assign more than one member to a task...

**Comment by krustymars** [comment]
I started to work on it, it's not a huge job, most of my time was consumed in figuring out the whole project architecture, where to look at, what change to be made at which place...

I modified the View for the creation, how does it feel to you ? 
![image](https://user-images.githubusercontent.com/2923696/29741111-8fd19200-8a66-11e7-8b2e-e891434f94e9.png)

I created a new Model class for the task-member n-to-n association, and routines I think will be needed

I've looked at how the DB schema migration works, it seems pretty easy to do it and avoid breaking things

I'm looking at which changes I have to perform in the task creation Controller to get the whole thing complete

This is for the 1st POC, allowing to give 1 owner and multiple assignees to a task at its creation


Afterwards, I'll look for modifying the task details and task dropdown menu to allow adding/removing members to an existing task

**Comment by asmwarrior** [comment]
Great work! Once ready, I'm grad to test it.

**Comment by poum** [comment]
Hi, i'm interested in this feature too. Regarding the screenshot, how will it possible to remove one assignee ? Maybe a list of assignees with an delete icon should be more adapted ?
Anyway, thanks for the job

**Comment by krustymars** [comment]
Okay, I have the basics features which seem to work (create/update/delete, and showing the multiple assignees in the task edit)

I'll try to run the unit test to see if they run fine before continuing.

Next step is mainly about the display of avatars in the boardview/listview, and in the task details.

**Suggestions would be welcome for the display of several avatars**, I'm not a graphic designer.

Once the feature complete and the tests pass I'll send a pull request with my code, those who are interested will be able to have it then.

**Comment by krustymars** [comment]
some eye-candies of my progress so far ;)

**Board view with all users avatars**
![image](https://user-images.githubusercontent.com/2923696/29754381-173ea8fe-8b84-11e7-9917-9c96ded65fb1.png)

**Creation of a new task with multiple assignees allowed**
![image](https://user-images.githubusercontent.com/2923696/29754419-b336158a-8b84-11e7-8ff5-da551e3bfae6.png)

**Modification of an existing multi-assignees task**
![image](https://user-images.githubusercontent.com/2923696/29754392-3f7cc710-8b84-11e7-9893-fed0d565ef76.png)

**Task details displaying mutliple assignees**
![image](https://user-images.githubusercontent.com/2923696/29754753-3899a474-8b8c-11e7-8262-23a9c3d2e235.png)

**Comment by rautamiekka** [comment]
^ Since the assignees has 'admin', the assignees are usernames instead of real names or some other names for the purpose of multi-assignee tasking ?

**Comment by krustymars** [comment]
@rautamiekka I gave the name "admin" to the admin user, who has "admin" username by default ;)

Assignees use the name, not the username, like in mono-user implementation.

**Comment by krustymars** [comment]
Additions 

Display of the multiple assignees in the task details view (same display for summary/activity stream/transitions/analytics)
![image](https://user-images.githubusercontent.com/2923696/29824066-c5a25f42-8cd0-11e7-8bc9-fea496f24627.png)

Display of the multiple assignees task list view
![image](https://user-images.githubusercontent.com/2923696/29824029-ababf2a6-8cd0-11e7-8ad3-2620b50b7eb3.png)

**QUESTION**
I feel like the avatars of assignees of a subtask shall appear as well in the task view, or shall be automatically assigned as the main task assignees (and thus their avatar will be displayed)

I'm not planning to integrate this in my current implementation of feature 85, but it may be a discussion to have for another feature request.
What do you think about this ?

**Comment by krustymars** [comment]
Implementation completed, unit test written and passed, code pushed, enjoy the feature !

You can try it as pull request #3428 on the kanboard official repository, or directly by cloning my fork of the repository at https://github.com/krustymars/kanboard

**Comment by asmwarrior** [comment]
One question is: Is there any method to migrate the database from the old official kanboard to the new "Multiple users in one task" branch? Thanks.

**Comment by krustymars** [comment]
The update of the schema is already included in my branch. 

From what I observed, the app apply schema updates at startup, so the procedure is to stop your current instance of kanboard, apply the branch, then restart your kanboard instance.

If you want to see the details of schema update I created it's in app/Schema/{$DBname}.PHP

Basically one more table task_has_users taking couples {task_id, user_id} as primary key

**Comment by tpimont** [comment]
Very good job krustymars .
I think many of us have waiting  a long time  this feature.

I have migrate my instance to the 85 banch .
The only annoying problem is that i have lost all the letters avatars of assigned task .
The only way is to reassigned the task.
May be it's a good thing to improve the database migration process to have a better experience .

**Comment by krustymars** [comment]
Hi @tpimont , thanks for your feedback :+1: 

To answer your request with a joke "this is not a problem, it's a feature!"

tldr: the owner name already appears on the task title bar, the assignees letters appear in circles

To give you more insights : 
With this implementation of the feature, 2 different notions appears now : the task owner (only one person, who was previously the person assigned to the task) and the task assignees.
In order to provide an implementation that fits the larger number of organizations use cases, I implemented it with the following specification in mind
- the owner role is person responsible for the task being executed (it doesn't imply actually executing it himself)
- the assignee role is the person actually executing the task (it doesn't imply he becomes the owner)
- a user can be both a owner and an assignee on a task (being both responsible for it, and actually executing it)
- a task can have a unique owner or no owner
- a task can have several assignees or no assignees

The use cases I intend to cover by doing so is:
- "managed team" team where there is one person responsible for the task execution (the owner), but is not involved in executing the task. The assignees are the people who execute the task.
- "self organizing team" team where all the assignees are responsible for the task execution and self organize to execute it.
- "leader organized team" team where the owner and the assignees contribute to the task execution. The owner can be seen as a leader or a senior contributing to the task as the assignees do, and being the responsible for its good execution.

From your feedback, I realize that it wouldn't be bad that the following assumption is made
'If a task has been assigned an owner and no assignee, the owner is the assignee'
Which will bring back the circled letters or avatars on your tasks where there is only one owner and no assignees, and avoid breaking previous behaviour on tasks with no assignees.

What do you think about it @tpimont ?
Other people, I'm also interested in your feedback please !

**Comment by tpimont** [comment]
Before migration 
![image](https://user-images.githubusercontent.com/25609323/30165819-b7e25ed0-93e1-11e7-8af7-749243034755.png)
After migration
![image](https://user-images.githubusercontent.com/25609323/30165859-dd2068d6-93e1-11e7-9af5-1df28837719e.png)

All avatars are lost even if owner is not the assigne .
The only way to bring back the avatar is to re assign the task  ( task 153 ).

So when you hava a lot of task reassign is not the solution and cant be done .

**Comment by rajo** [comment]
@krustymars with your final assumption about owner and no assignee (in order to avoid breaking changes) I'd say this would be very useful then

**Comment by krustymars** [comment]
@tpimont 
yes I see

What do you think of the proposition made in my previous message

> From your feedback, I realize that it wouldn't be bad that the following assumption is made
"If a task has been assigned to an owner and has no assignee assigned, the owner becomes the assignee by default"
Which will bring back the circled letters or avatars on your tasks where there is only one owner and no assignees, and avoid breaking previous behaviour on tasks with no assignees.

Or simpler, is there is no assignees and a owner, the task displays the avatar of the owner. This would avoid you to reassign all your tasks

**Comment by tpimont** [comment]
Yes but as you can see we have task that are not assign to the owner and the avatar is also gone.
So what can be done for this tasks ?
Can a sql scipt be apply to the database to migrate data of assigne to the new format ?

**Comment by krustymars** [comment]
@tpimont @rajo 
Yes basically the 2 options here are

Option 1 : modifying only the display, in order that a task with only a owner and no assignees will display the avatar of the owner. If assignees are added to this task, the avatar of the owner will disappear, unless he is also added to assignees of the task.

Option 2: 
- assigning the owner as assignee by default to any new task, if no assignee has been selected in the assignee list. The user being both the owner and the assignee, is avatar will appear.
- modifying the database update when the branch of the feature is applied, in order that any already existing task having a owner and no assignee will be added the owner as an assignee. The avatar of the owner will then appear, because he have became an assignee in the database

I like option 1 very much because it's less work to do :laughing:

**Comment by tpimont** [comment]
Option 3:

modifying the database update when the branch of the feature is applied, in order that any already existing task having  assignee became an assignee in the database
So the previous avatar appear even if it is the owner avatar .

If there is no asigne let the avatar blank.

**Comment by krustymars** [comment]
@tpimont indeed
I let a little bit more time to other users to give their feedback or other propositions

**Comment by RussellAult** [comment]
If I can make a suggestion, my sense is that the least-breaking implementation would be to give Project Managers the option of whether to allow for multiple assignees on a project-by-project basis:

Add a new project setting "Allow multiple assignees", with the default being false. While false, the current behaviour and appearance is maintained, with a single assignee being stored as the "owner". While true, the current "Assignee" drop-down is replaced by some form of multi-select, using a new data structure to store the multiple assignees; in this mode, the "owner" is completely ignored.

When the project setting is switched from false to true, the owner of each task in the project is added as an entry for that task into the new assignees data structure. When the project setting is switched from true to false any task with only one assignee has that user set as the task's owner, any task with more than one assignee will have its owner set as undefined, and a warning confirmation dialogue pops up explaining what's about to happen.

Because "Allow multiple assignees" is false by default, upgrading causes no immediate functionality change, and changes only start happening when the setting is enabled for a particular project.

**Comment by krustymars** [comment]
@russellault
Thanks for your relevant feedback. Indeed, what you're proposing would be perfect, though it's a lot more work to be performed before all the users of kanboard can benefit from the multiple users in a task feature. I can totally see your proposition as the final result of the feature implementation.

However in the meantime, a lot of kanboard users would like to have this feature, at least minimally implemented. It's a kind of balance between heaving to wait, and having the final perfect version of the feature. I think @fguillot might be the most appropriate person to say which balance would be desirable.

In my opinion, the proposition made by @tpimont is the more adequate as it avoids breaking anything, and let the feature be available for all users wanting to have it, and not involving a whole lot of additional development for me, my time to do it is scares at the moment.

What do you think @fguillot ?

**Comment by tpimont** [comment]
Hope this could be in the next 1.0.47 release !
I can test the new migration scheme ( not breaking anything ) if you want .

**Comment by x9t9** [comment]
This is a very welcomed ( and long-waited ) feature .

In relation to the "owner and no assignee"  discussion above - In real world scenario, a task should not be created if there is no one to assign it to . If an assignee has not yet been selected, then the assignee should be the project manager ( or equivalent ) and the task is to find an assignee :-) 

In that regards I suggest that the creator will be assignee by default , or ( more complicated code - wise ) have a default assignee ( project manager ) for each project .  

Code wise - I would use the [select2 ](https://github.com/select2/select2 ) script that is already implemented on kanboard and included in core->vendor->select2
You can see it in action with search capabilities at the top bar ( e.g. "select another project" ) .

**Comment by krustymars** [comment]
> In real world scenario, a task should not be created if there is no one to assign it to 

I see a lot of organizations (including mine) where tasks are created in bulk far ahead knowing who is going to execute them

> If an assignee has not yet been selected, then the assignee should be the project manager ( or equivalent ) and the task is to find an assignee :-)

This would break functionality for people who automatize task coloration based on no one being allocated to it, for filtering tasks with no user allocated, and so on... The assumption "the project manager is the assignee until one assignee is assigned" doesn't hold for all organizations IMHO

However no one forbids you to allocate the project manager to your newly created tasks when they are created, and then the user that finally will execute them. ;)

**Comment by x9t9** [comment]
> The assumption "the project manager is the assignee until one assignee is assigned" doesn't hold for all organizations IMHO
**TLDR ;** 
It should :-) - at least in the correct work methodology . 
But I do understand that Kanban leaves a lot of flexibility space for "each to have it's own" - and that's ok .. 

Still, a task without assignee will never be completed , and therefore is not a real task - until an assignee is assigned . This is the job of the Kanban-Master ( Product owner to use thee real term , or  project manager ) - and one of his jobs is to keep the kanboard clean and only with relevant tasks for thee team to execute . 
In Kanban terms , the "product owner" has a very important role , and some of the important tasks/goals for him would be :

1 - Matching the amount of work in progress (WIP) to the team's capacity.
2- Short Cycle Time ( it's a key metrics )
3 - Fast identification of blockers and dependencies
( and 4 - CD / CI  where implemented ..  ) 

In the view of the above points , a task without an assignee conflicts with all of them  - and therefor while not an ABSOLUTE wrong ( no such thing .. ) , it is not conforms to the "pure" kanban way .

But , like above stated , I do appreciate that the beauty of kanban is that it leaves a lot of flexibility and maneuver space for each to adapt their own work methodologies , and I do not want to turn this into a discussion about _kanban_ itself ( although relevant in the sense of how to implement features ) .

If you need help in implementing the UI features  ( and DB without conflicts or migration ) then let me know ..

**Comment by piltraspn** [comment]
Hi everyone!
             Sorry to ask, I don't really know the steps here but... any idea when will this feature be available?... It's been more than 3 years, I see people worked very hard doing the most difficult part with an awesome result but still not available when in my case this feature is absolutely necessary. Without it 90% of Kanboard doesn't make sense to my purposes. I would really prefer using Kanboard than any other alternative though!
    Any clues?
Thank you so much for this great feature!

**Comment by krustymars** [comment]
@piltraspn
You can use it from my pull request if you don't want to wait for integration in the master

https ://github.com/kanboard/kanboard/pull/3431

**Comment by krustymars** [comment]
@dixon22ma  indeed I didn't realize that a merge have been tryed in the main branch.
I just checked, the error is because some other PR involving an update of the DB scheme have been merged before mine, involving the DB version being updated. 
So I have to update the DB version key in my implementation, hopefully no big deal here. I'll do it as soon as I have a little slot.

**Comment by dixon22ma** [comment]
@krustymars we will wait for your update greetings ...

**Comment by creecros** [comment]
For those interested, https://github.com/creecros/group_assign

**Comment by mapeze** [comment]
After successful install of the plugin, an improvement proposal:

The task, if assigned to a group, it should be visible to all group members in their calendar view.

**Comment by creecros** [comment]
I've never touched the calendar, so I'll have to do some research on that one.

**Comment by lastlink** [comment]
@mapeze you should give feature requests for plugins directly on their git repos rather than here. https://github.com/creecros/group_assign/issues. @creecros plugin looks great. I think we need an additional assignees plugin similar to the 85 that will have a multiselect like box w/ no limit to users & separate from main assignee. This will avoid database breaks by having it separate.

**Comment by creecros** [comment]
@lastlink the 85? I'm not sure i'm following you.

**Comment by creecros** [comment]
Oh, Feature 85....I'm slow to catch up...hell i just realized this wasn't group_assign issue page.

**Comment by creecros** [comment]
@lastlink off the top of my head I think the plug-in would be easily modifiable to accommodate that but in its current state you would not be able to use both plugins at the same time...overrides...I'd have to figure out how remove those

**Comment by mapeze** [comment]
I opened an issue on the git repos. Not sure if I did it as I should do.

**Comment by lastlink** [comment]
@creecros I think the plugin I mentioned would solve this issue, the group assigned to task would not make this issue closed. I agree that the group plugin as well as the pull request would be good starters to develop the multi assigned user plugin in a separate text input field from owner. I don't have a pushing need for this right now. I may end up eventually developing it if somebody else doesn't.

**Comment by seancarson** [comment]
Test 22

**Comment by tuxayo** [comment]
If someone wants to work on this, it seems that #3431 is a good base (needs migration for existing data + maybe rebasing)

**Comment by creecros** [comment]
@lastlink 

https://github.com/creecros/Group_assign/releases/tag/1.1.0

Added a multiselect for additional assignees, which I labeled `Other Assignees` 

![image](https://user-images.githubusercontent.com/26339368/49529059-3b0e5780-f883-11e8-9d4f-9a85691cd34c.png)

![image](https://user-images.githubusercontent.com/26339368/49529021-29c54b00-f883-11e8-99c4-936611ab72e6.png)

Also, fixed previous filter, it will now find `assignee:name` inside the group or multiselect and return the task correct tasks with assignees.

**Comment by bugwatcher** [comment]
Please, use the [forum](https://kanboard.discourse.group) (https://kanboard.discourse.group) for discussions and feature requests, the bug tracker is reserved only for bugs.
